### ISSUE TITLE
Temporality [feature request]

### ISSUE THREAD

**Issue by snosrap** [medspacy/medspacy#26]
Hi MedSpaCy,

Thanks for creating such a great library -- I've found it incredibly helpful!

I wonder if you have any thoughts on creating a "temporality" component. That is, for a given entity, when did that entity start happening? This would be useful for symptoms in an HPI (e.g., "Pt initially presented with 2 weeks of productive cough and developed fever 3 days ago.") Symptom duration might give a sense of where a patient is in the course of an illness (e.g., COVID).

I've been developing a naïve approach of a `TemporalityComponent`, which is an extension on a `Span` that just looks for the nearest `Span` labelled as a "DATE". SpaCy's native models (e.g., "en_core_web_sm") label "DATE"s out-of-the-box so it's an easy implementation.

Curious to know if you've considered/evaluated a more sophisticated rules-based approach.

Thanks!

**Comment by abchapman93** [comment]
Thanks for using medspaCy and sharing your experience! Love the idea of a temporality component. I don't think any of us have explicitly done what you're describing or have plans in the near future, but I think it would be useful generally. We do handle some temporality using ConText and the `is_historical` attribute, although that's slightly different from what you're doing here.

An alternative to adding a new component is to handle this using ConText. You can write `ConText` rules with a category of `"TEMPORAL"` which will connect the temporal expressions to the entities in the text. This will have the benefit of letting you control which direction to look in the sentence. Here's an example using both the NER-based approach and using a custom rule:

```python
import spacy
from spacy.tokens import Span
from medspacy.context import ConTextComponent, ConTextItem

nlp = spacy.load("en_core_web_sm")

texts = [
    "Dx COVID on 10/11/2020.",
    "Dx COVID 2 weeks ago.",
]

docs = list(nlp.pipe(texts))

context = ConTextComponent(nlp)

context.add([ 
    # Based on out-of-the-box spaCy NER DATE
    ConTextItem("DATE", "TEMPORAL", rule="BIDIRECTIONAL", pattern=[{"ENT_TYPE": "DATE"}]),
    
    # Rule-based expression
    ConTextItem("<D> days ago", "TEMPORAL", rule="BACKWARD", pattern=[
        {"LIKE_NUM": True},
        {"LEMMA": {"IN": ["day", "week", "month", "year"]}},
        {"LOWER": "ago"}
    ]),
])

# Manually set entity spans and run context
for doc in docs:
    doc.ents += (Span(doc, 1, 2, "CONDITION"),)
    context(doc)
    
# Print out temporal connections
for doc in docs:
    for (ent, modifier) in doc._.context_graph.edges:
        if modifier.category == "TEMPORAL":
            print(ent, " --> ", modifier.span)
```

It would be great to add some "TEMPORAL" rules into the ConText knowledge base, so if you develop any before we get to it we'd love it if you could share them with us! Some resources that could be helpful for this could be [Timex](http://www.timeml.org/tempeval2/tempeval2-trial/guidelines/timex3guidelines-072009.pdf) or a model trained on the [i2b2 2012 temporal expressions challenge](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3756273/).

**Comment by snosrap** [comment]
Thanks!
I've only used medspacy.context in its "out-of-the-box" configuration -- I didn't realize it could be configured/extended to support this use case. I'll let you know if/when I come up with anything.
Thanks also for the pointer to the n2c2/i2b2 2012 dataset -- this is exactly what I needed to test some of my assumptions without having to annotate a bunch of notes. Turns out if you filter out the references to "note section" dates, the heuristic we're talking about (nearest DATE to an ENTITY) works pretty well.
Thanks again for your thoughtful suggestions!
Ford
